### PR TITLE
Add models for OAuth2, Person, ServiceAccount and add missing endpoints

### PR DESCRIPTION
**Pull request by sbordeyne**
Fixes #

Checklist

- [x] This pr contains no AI generated code
- [ ] cargo fmt has been run
- [ ] cargo clippy has been run
- [ ] cargo test has been run and passes
- [ ] book chapter included (if relevant)
- [ ] design document included (if relevant)

### CODE REVIEW

**Comment by sbordeyne** [comment]
@yaleman here is the PR for the models, if you want to merge it into your branch

File: pykanidm/kanidm/models/group.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        # we want either the first element of gidnumber_field, or None
+        gidnumber_field = self.attrs.get("gidnumber", [])
+        gidnumber: Optional[int] = None
+        if len(gidnumber_field) > 0:
+            gidnumber = int(gidnumber_field[0])
+
+        return Group(
+            name=self.attrs["name"][0],
+            uuid=self.attrs["uuid"][0],
+            spn=self.attrs["spn"][0],
```
Comment (yaleman) [inline pykanidm/kanidm/models/group.py:50]: Can you please ensure you're checking the length up in the "for field in ..." section?  It'll throw an ick error here (also ensure this is done in the other classes)

File: pykanidm/kanidm/types.py [hunk 2]
```diff
@@ -4,14 +4,19 @@
...
 
-from pydantic import field_validator, ConfigDict, BaseModel, Field, RootModel
+from pydantic import field_validator, ConfigDict, BaseModel, Field
 import toml
 
+from kanidm.models.group import Group as GroupInfo, RawGroup as RawGroupInfo, GroupList  # noqa
 
-class ClientResponse(BaseModel):
+
+T = TypeVar("T", Dict[str, Any], None)
```
Comment (yaleman) [inline pykanidm/kanidm/types.py]: This is not just a Dict[str, Any] = it can also be a string / list / etc - it's JSON-decoded from the payload.

**Review by yaleman — CHANGES_REQUESTED** [review]
Where you replaced the `ClientResponse` with bare responses, you're assuming that it'll always work, which isn't valid. Please restore those change as you've changed the response types and assumptions around success of the calls.

**Comment by yaleman** [comment]
And please ensure the tests pass - mypy's screaming.

**Comment by sbordeyne** [comment]
> Where you replaced the `ClientResponse` with bare responses, you're assuming that it'll always work, which isn't valid. Please restore those change as you've changed the response types and assumptions around success of the calls.

What I did is type the field data according to what a successful response will be for the fields, the methods still return a bare client response, it's just more accurate typing-wise (by using a Generic in conjunction with pydantic).

In my opinion, this change is better than typing with an Any, because as a user of this library I didn't know what to expect from the bare client response's data field.

Side note : I bumped the semver of kanidm to 1.0.0, since there is technically a breaking change (the Endpoints class instead of the dict). Due to this I could also remove the aliasing from the types module for the group models. Let me know what you prefer : either I add the dict back (and move on to 0.1.0) with a bit of a dirty hack, or I remove the model imports from the types module.

File: pykanidm/kanidm/__init__.py [hunk 3]
```diff
@@ -489,22 +499,115 @@ async def get_radius_token(self, username: str) -> ClientResponse:
...
+    async def oauth2_rs_get(self, rs_name: str) -> OAuth2Rs:
+        """get an OAuth2 client"""
+        endpoint = f"{Endpoints.OAUTH2}/{rs_name}"
+        response: ClientResponse[IOauth2Rs] = await self.call_get(endpoint)
+        if response.status_code != 200:
+            raise ValueError(f"Failed to get oauth2 resource server: {response.content}")
+        if response.data is None:
+            raise ValueError(f"Failed to get oauth2 resource server: {response.content}")
+        data = RawOAuth2Rs(**response.data).as_oauth2_rs
+        secret = await self.oauth2_rs_secret_get(rs_name)
```
Comment (yaleman) [inline pykanidm/kanidm/__init__.py]: Do NOT do this. It's not our API.

**Review by yaleman — CHANGES_REQUESTED** [review]
First, please don't break our security designs by making changes to the OAuth2 API.

Returning a typed `ClientResponse` is fine, but when I made the original comment, you'd replaced it with the inner type which didn't match what would actually come out.

The tests are still failing (please run `make test/pykanidm` before committing again). You've got a load of duplicated function defs in there amongst other things.

```
Running mypy
kanidm/types.py:44: error: Missing type parameters for generic type "ClientResponse"  [type-arg]
kanidm/types.py:64: error: Missing type parameters for generic type "ClientResponse"  [type-arg]
kanidm/types.py:78: error: Missing type parameters for generic type "ClientResponse"  [type-arg]
kanidm/__init__.py:340: error: Missing type parameters for generic type "dict"  [type-arg]
kanidm/__init__.py:501: error: Missing type parameters for generic type "ClientResponse"  [type-arg]
kanidm/__init__.py:901: error: Returning Any from function declared to return "list[str]"  [no-any-return]
kanidm/__init__.py:932: error: Name "oauth2_rs_list" already defined on line 511  [no-redef]
kanidm/__init__.py:932: error: Missing type parameters for generic type "ClientResponse"  [type-arg]
kanidm/__init__.py:939: error: Name "oauth2_rs_get" already defined on line 523  [no-redef]
kanidm/__init__.py:939: error: Missing type parameters for generic type "ClientResponse"  [type-arg]
kanidm/__init__.py:945: error: Name "oauth2_rs_delete" already defined on line 550  [no-redef]
kanidm/__init__.py:945: error: Missing type parameters for generic type "ClientResponse"  [type-arg]
kanidm/__init__.py:951: error: Name "oauth2_rs_basic_create" already defined on line 556  [no-redef]
kanidm/__init__.py:953: error: Missing type parameters for generic type "ClientResponse"  [type-arg]
kanidm/__init__.py:968: error: Missing type parameters for generic type "ClientResponse"  [type-arg]
kanidm/__init__.py:974: error: Name "_validate_is_valid_origin_url" already defined on line 573  [no-redef]
kanidm/__init__.py:989: error: Name "oauth2_rs_public_create" already defined on line 588  [no-redef]
kanidm/__init__.py:991: error: Missing type parameters for generic type "ClientResponse"  [type-arg]
tests/test_oauth2.py:33: error: "list[OAuth2Rs]" has no attribute "data"  [attr-defined]
tests/test_oauth2.py:34: error: "list[OAuth2Rs]" has no attribute "status_code"  [attr-defined]
tests/test_oauth2.py:36: error: "list[OAuth2Rs]" has no attribute "data"  [attr-defined]
tests/test_oauth2.py:37: error: "list[OAuth2Rs]" has no attribute "data"  [attr-defined]
Found 22 errors in 3 files (checked 26 source files)
make: *** [test/pykanidm/mypy] Error 1
```

lint errors:
```
 ❯ make test/pykanidm/lint
cd pykanidm && \
	poetry install && \
	poetry run ruff tests kanidm
Installing dependencies from lock file

No dependencies to install or update

Installing the current project: kanidm (1.0.0)
kanidm/__init__.py:932:15: F811 Redefinition of unused `oauth2_rs_list` from line 511
kanidm/__init__.py:939:15: F811 Redefinition of unused `oauth2_rs_get` from line 523
kanidm/__init__.py:945:15: F811 Redefinition of unused `oauth2_rs_delete` from line 550
kanidm/__init__.py:951:15: F811 Redefinition of unused `oauth2_rs_basic_create` from line 556
kanidm/__init__.py:989:15: F811 Redefinition of unused `oauth2_rs_public_create` from line 588
kanidm/models/service_account.py:28:9: F841 Local variable `required_fields` is assigned to but never used
kanidm/types.py:7:65: F401 [*] `typing.TypedDict` imported but unused
Found 7 errors.
[*] 1 fixable with the `--fix` option (1 hidden fix can be enabled with the `--unsafe-fixes` option).
make: *** [test/pykanidm/lint] Error 1
```

**Comment by sbordeyne** [comment]
@yaleman I'll fix this up sometime during the following week (I've been quite busy with work and starting a business hence why it took a bit of time) Let me know if there's any other comments. Duplicated function definitions happened because I botched the git merge on my branch, my bad.

**Comment by sbordeyne** [comment]
@yaleman Sorry for the late input here, I've fixed all of the mypy errors, I'm still fixing the tests though. Any reason why the testing fixtures use an actual live kanidm config file for testing purposes ? They fail on my machine (thank god as well that they fail) because my kanidm config (actual live config for my server) is wrong. For a piece of software whose sole purpose is security, it's quite negligent to use a prod server for testing. 

Good practice would be to host an actual staging env to run the test suite against, and pass the config as environment variables (so that you can set up secrets in github for that server) instead of reading `~/.config/kanidm`.

I'll forcefully skip the tests to fix the remaining, but if any of them fail, I won't be able to fix them. I don't feel comfortable using my own instance for testing purposes.

**Comment by yaleman** [comment]
> @yaleman Sorry for the late input here, I've fixed all of the mypy errors, I'm still fixing the tests though. Any reason why the testing fixtures use an actual live kanidm config file for testing purposes ? They fail on my machine (thank god as well that they fail) because my kanidm config (actual live config for my server) is wrong. For a piece of software whose sole purpose is security, it's quite negligent to use a prod server for testing.

It's not *intended* for devs to be using a prod server for testing, and yes, we could definitely improve the documentation around testing - but ... any Kanidm server which is running is _effectively_ live, so how would you delineate the two? I'm happy to accept constructive suggestions.

> Good practice would be to host an actual staging env to run the test suite against, and pass the config as environment variables (so that you can set up secrets in github for that server) instead of reading `~/.config/kanidm`.

The project doesn't have capability for, nor should it be expected to host testing infra... this is something that can be spun up in ~30 seconds with docker. 

`./scripts/setup_dev_environment.sh` can do all the config for a local server and provide creds. As noted previously, the Python lib's in very early development stages.

> I'll forcefully skip the tests to fix the remaining, but if any of them fail, I won't be able to fix them. I don't feel comfortable using my own instance for testing purposes.

It's specifically called out in the README for the package - either way you'd need to have set it up specifically for testing - or you're welcome to skip the live tests.

> Pytest it used for testing, if you don't have a live server to test against and config set up, use
> `poetry run pytest -m 'not network'`.

**Comment by sbordeyne** [comment]
> It's not intended for devs to be using a prod server for testing, and yes, we could definitely improve the documentation around testing - but ... any Kanidm server which is running is effectively live, so how would you delineate the two? I'm happy to accept constructive suggestions.

True, in the sense that any running kanidm server is technically live, but the main issue is that the fixtures actually try to read the config at ~/.config/kanidm. That's the root of the problem. Since kanidm can be spun up using docker, the make target could :

- Run kanidm in docker first
- Write the credentials for kanidm (certificates, config) in /tmp, or in the repo (adding the files to .gitignore as required)
- Seed the database with test data (when applicable)
- Run the test suite against the dev environment
- Trash the container

> ./scripts/setup_dev_environment.sh can do all the config for a local server and provide creds. As noted previously, the Python lib's in very early development stages.

this script requires ripgrep to be installed and in the path, it would be nice to have a script to install dependencies, or just rely on common GNU/bash utils that are present everywhere

**Comment by sbordeyne** [comment]
Tests are still breaking on my end, I'm just committing to not lose progress

**Comment by yaleman** [comment]
> Tests are still breaking on my end, I'm just committing to not lose progress

No worries, will wait for you to flag it as ready to check.

**Comment by yaleman** [comment]
> > It's not intended for devs to be using a prod server for testing, and yes, we could definitely improve the documentation around testing - but ... any Kanidm server which is running is effectively live, so how would you delineate the two? I'm happy to accept constructive suggestions.
> 
> True, in the sense that any running kanidm server is technically live, but the main issue is that the fixtures actually try to read the config at ~/.config/kanidm. That's the root of the problem. Since kanidm can be spun up using docker, the make target could :
> 
> * Run kanidm in docker first
> * Write the credentials for kanidm (certificates, config) in /tmp, or in the repo (adding the files to .gitignore as required)
> * Seed the database with test data (when applicable)
> * Run the test suite against the dev environment
> * Trash the container
>
> > ./scripts/setup_dev_environment.sh can do all the config for a local server and provide creds. As noted previously, the Python lib's in very early development stages.
> 
> this script requires ripgrep to be installed and in the path, it would be nice to have a script to install dependencies, or just rely on common GNU/bash utils that are present everywhere

This PR started from a request for a few endpoints which turned into *much* more work than was originally requested and I'm trying to limit the scope blowing out further. I'm not going to get into rebuilding the test scripts in this discussion. I get it's not optimal, but we're an incredibly resource-constrained project, and as mentioned previously the Python code's done as a best-effort thing.

We'll happily accept a PR to `kanidm/kanidm` with those suggestions (a separate one, *please*)  if you want to implement them, or a new issue requesting the work be done and it'll be triaged appropriately.

**Comment by sbordeyne** [comment]
> This PR started from a request for a few endpoints which turned into much more work than was originally requested and I'm trying to limit the scope blowing out further. I'm not going to get into rebuilding the test scripts in this discussion. I get it's not optimal, but we're an incredibly resource-constrained project, and as mentioned previously the Python code's done as a best-effort thing.

> We'll happily accept a PR to kanidm/kanidm with those suggestions (a separate one, please) if you want to implement them, or a new issue requesting the work be done and it'll be triaged appropriately.

yes I completely understand, it blew out already quite a bit (for the better in my opinion though). That's why I'm trying my best to fix the tests and make them work on my machine. AFAIK the only breaking test is the oauth2_rs one, which fails because I changed the client to return parsed models (and not the raw ClientResponse) for usability reasons.

I'll get it done during the week (hopefully, if my full time job leaves me enough energy to work on it at night)

**Pull request merged**